arcgis-ios-sdk-samples/Maps/Generate offline map/GenerateOfflineMapViewController.swift

```diff
@@ -190,91 +169,135 @@ class GenerateOfflineMapViewController: UIViewController, AGSAuthenticationManag
-        let yesAction = UIAlertAction(title: "Yes", style: .default) { [weak self] (action) in
+        let alertController = UIAlertController(title: nil, message: "This sample requires you to login in order to take the map's basemap offline. Would you like to continue?", preferredStyle: .alert)
+        let yesAction = UIAlertAction(title: "OK", style: .default) { [weak self] (action) in
             self?.addMap()
         }
         
-        let noAction = UIAlertAction(title: "No", style: .cancel)
+        let noAction = UIAlertAction(title: "Cancel", style: .cancel)
         
         alertController.addAction(noAction)
         alertController.addAction(yesAction)
         
-        self.present(alertController, animated: true, completion: nil)
+        present(alertController, animated: true, completion: nil)
     }
     
-    func frameToExtent() -> AGSEnvelope {
-        let frame = self.mapView.convert(self.extentView.frame, from: self.view)
+    private func extentViewFrameToEnvelope() -> AGSEnvelope {
+        
+        let frame = mapView.convert(extentView.frame, from: view)
+        
+        //the lower-left corner
+        let minPoint = mapView.screen(toLocation: frame.origin)
+        
+        //the upper-right corner
+        let maxPoint = mapView.screen(toLocation: CGPoint(x: frame.maxX, y: frame.maxY))
+        
+        //return the envenlope covering the entire extent frame
+        return AGSEnvelope(min: minPoint, max: maxPoint)
+    }
+    
+    private func getNewOfflineGeodatabaseURL()->URL{
+
+        //get a suitable directory to place files
+        let documentDirectoryURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
+        
+        //create a unique name for the geodatabase based on current timestamp
+        let formattedDate = ISO8601DateFormatter().string(from: Date())
         
-        let minPoint = self.mapView.screen(toLocation: frame.origin)
-        let maxPoint = self.mapView.screen(toLocation: CGPoint(x: frame.origin.x+frame.width, y: frame.origin.y+frame.height))
-        let extent = AGSEnvelope(min: minPoint, max: maxPoint)
-        return extent
+        return documentDirectoryURL.appendingPathComponent("\(formattedDate).geodatabase")
     }
 
     deinit {
         
-        guard let progress = self.generateOfflineMapJob?.progress else {
+        guard let generateOfflineMapJob = generateOfflineMapJob else {
             return
         }
+        let progress = generateOfflineMapJob.progress
         
         let isCompleted = (progress.totalUnitCount == progress.completedUnitCount)
         let isCancelled = progress.isCancelled
         
         if !isCancelled && !isCompleted {
             //remove observer
-            self.generateOfflineMapJob?.progress.removeObserver(self, forKeyPath: #keyPath(Progress.fractionCompleted))
+            generateOfflineMapJob.progress.removeObserver(self, forKeyPath: #keyPath(Progress.fractionCompleted))
```

Comment:
You already have a reference to the progress, so use that instead of going through the job.